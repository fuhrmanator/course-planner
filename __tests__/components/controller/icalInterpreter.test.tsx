import { describe, expect, test } from '@jest/globals';
import * as fs from 'fs';
import path from 'path';
import { parseICALEvents } from '@/components/controller/util/icalInterpreter';
import { CourseEvent, EventType } from '@/components/model/interfaces/courseEvent';
const ical = require('ical');
const dataPath = "__tests__/data";

function etsSample1() {
  // create an empty file blah.txt in the dataPath directory
  console.log(`Absolute path: ${path.resolve(path.join(dataPath, "blah.txt"))}`);

  // read a file ETS_SAMPLE_1.ics from disk and return its content as a string
  return fs.readFileSync(path.join(dataPath, "ETS_SAMPLE_1.ics")).toString();
}

const icalToEventField: { [key: string]: string } = {
  "start": "start",
  "end": "end",
  "uid": "uid",
  "summary": "title"
}

async function getFileContent(filename: string): Promise<string> {
  const data = await fs.promises.readFile(path.join(dataPath, filename));
  return data.toString();
}

function parseIcal(icalData: string): any {
  return ical.parseICS(icalData);
}

function eventEqualExpected(expectedResult: any, toTest: CourseEvent): boolean {
  let isEqual = true;
  for (let expectedField in icalToEventField) {
    // @ts-ignore
    isEqual = isEqual && JSON.stringify(toTest[icalToEventField[expectedField]]) === JSON.stringify(expectedResult[expectedField]);

  }

  return isEqual;
}

function generateUUID() {
  let uuid = '';
  for (let i = 0; i < 32; i++) {
    const randomDigit = Math.floor(Math.random() * 16);
    uuid += randomDigit.toString(16);
  }
  return uuid;
}

describe('ICAL module', () => {
  let icsData: string;
  let icalEvents: any;
  beforeEach(async () => {
    icsData = await getFileContent("valid.ics");
    icalEvents = parseIcal(icsData);
  })

  test('ICAL property are translated to CalEvents', async () => {
    // const icsData = await getFileContent("valid.ics");
    const parsedEvents = parseICALEvents(icsData);

    for (let parsedEvent of parsedEvents) {
      expect(parsedEvent.uid in icalEvents).toBeTruthy();
      expect(eventEqualExpected(icalEvents[parsedEvent.uid], parsedEvent)).toBeTruthy();
    }
  });
  test('Parser does not create events with unsupported types', async () => {
    const typeAttribute = "CATEGORIES:";
    const bogusTypeAttribute = typeAttribute + generateUUID();
    const eventNb = Object.keys(icalEvents).length;
    const bogusNb = Math.floor(eventNb / 2);
    let count = 0;
    const dataWithBogusTypes = icsData.replace(new RegExp(typeAttribute + ".*", "g"), (match, _) => {
      count++;
      return count <= bogusNb ? bogusTypeAttribute : match;
    });
    icalEvents = parseICALEvents(dataWithBogusTypes);
    expect(icalEvents.length).toBe(eventNb - bogusNb);
  });

  test('should parse iCal data into an array of CourseEvent objects', () => {
    const result = parseICALEvents(icsData);
    expect(result).toMatchSnapshot();
  });

});
