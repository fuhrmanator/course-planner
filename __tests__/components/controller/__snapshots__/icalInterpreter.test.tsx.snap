// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICAL module should parse iCal data into an array of CourseEvent objects 1`] = `
[
  {
    "dsl": [],
    "end": 2022-09-07T16:00:00.000Z,
    "start": 2022-09-07T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345453-1",
  },
  {
    "dsl": [],
    "end": 2022-09-14T16:00:00.000Z,
    "start": 2022-09-14T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345454-1",
  },
  {
    "dsl": [],
    "end": 2022-09-21T16:00:00.000Z,
    "start": 2022-09-21T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345455-1",
  },
  {
    "dsl": [],
    "end": 2022-09-28T16:00:00.000Z,
    "start": 2022-09-28T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345456-1",
  },
  {
    "dsl": [],
    "end": 2022-10-05T16:00:00.000Z,
    "start": 2022-10-05T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345457-1",
  },
  {
    "dsl": [],
    "end": 2022-10-12T16:00:00.000Z,
    "start": 2022-10-12T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345458-1",
  },
  {
    "dsl": [],
    "end": 2022-10-19T16:00:00.000Z,
    "start": 2022-10-19T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345459-1",
  },
  {
    "dsl": [],
    "end": 2022-10-26T16:00:00.000Z,
    "start": 2022-10-26T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345460-1",
  },
  {
    "dsl": [],
    "end": 2022-11-02T16:00:00.000Z,
    "start": 2022-11-02T12:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345461-1",
  },
  {
    "dsl": [],
    "end": 2022-11-16T17:00:00.000Z,
    "start": 2022-11-16T13:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345462-1",
  },
  {
    "dsl": [],
    "end": 2022-11-23T17:00:00.000Z,
    "start": 2022-11-23T13:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345463-1",
  },
  {
    "dsl": [],
    "end": 2022-11-30T17:00:00.000Z,
    "start": 2022-11-30T13:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345464-1",
  },
  {
    "dsl": [],
    "end": 2022-12-07T17:00:00.000Z,
    "start": 2022-12-07T13:30:00.000Z,
    "title": "LOG121-01 (C)",
    "type": 0,
    "uid": "http://signets.etsmtl.ca/horaire/345465-1",
  },
  {
    "dsl": [],
    "end": 2022-09-12T20:30:00.000Z,
    "start": 2022-09-12T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345440-1",
  },
  {
    "dsl": [],
    "end": 2022-09-19T20:30:00.000Z,
    "start": 2022-09-19T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345441-1",
  },
  {
    "dsl": [],
    "end": 2022-09-26T20:30:00.000Z,
    "start": 2022-09-26T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345442-1",
  },
  {
    "dsl": [],
    "end": 2022-10-11T20:30:00.000Z,
    "start": 2022-10-11T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345444-1",
  },
  {
    "dsl": [],
    "end": 2022-10-17T20:30:00.000Z,
    "start": 2022-10-17T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345445-1",
  },
  {
    "dsl": [],
    "end": 2022-10-24T20:30:00.000Z,
    "start": 2022-10-24T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345446-1",
  },
  {
    "dsl": [],
    "end": 2022-10-31T20:30:00.000Z,
    "start": 2022-10-31T17:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345447-1",
  },
  {
    "dsl": [],
    "end": 2022-11-07T21:30:00.000Z,
    "start": 2022-11-07T18:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345448-1",
  },
  {
    "dsl": [],
    "end": 2022-11-14T21:30:00.000Z,
    "start": 2022-11-14T18:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345449-1",
  },
  {
    "dsl": [],
    "end": 2022-11-21T21:30:00.000Z,
    "start": 2022-11-21T18:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345450-1",
  },
  {
    "dsl": [],
    "end": 2022-11-28T21:30:00.000Z,
    "start": 2022-11-28T18:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345451-1",
  },
  {
    "dsl": [],
    "end": 2022-12-05T21:30:00.000Z,
    "start": 2022-12-05T18:30:00.000Z,
    "title": "LOG121-01 (Labo)",
    "type": 1,
    "uid": "http://signets.etsmtl.ca/horaire/345452-1",
  },
]
`;
